client/src/game/reducers/noteIdentity.ts

```diff
@@ -47,91 +47,88 @@ export function parseIdentity(variant: Variant, keyword: string): CardIdentity {
   return { suitIndex, rank };
 }
 
-function possibilitiesComplement(
-  variant: Variant,
-  possibilities: Array<[number, number]>,
-): Array<[number, number]> {
-  const complement: Array<[number, number]> = [];
-  for (let suitIndex = 0; suitIndex < variant.suits.length; suitIndex++) {
-    for (const rank of variant.ranks) {
-      // Ensure that this possibility is not present in the source list
-      if (
-        !possibilities.some(
-          ([negatedSuitIndex, negatedRank]) =>
-            negatedSuitIndex === suitIndex && negatedRank === rank,
-        )
-      ) {
-        complement.push([suitIndex, rank]);
+function identityMapToArray(cardMap: number[][]): Array<[number, number]> {
+  const rows = cardMap.length;
+  const cols = cardMap[0].length;
+  const possibilities: Array<[number, number]> = [];
+  for (let row = 0; row < rows; ++row) {
+    for (let col = 0; col < cols; ++col) {
+      if (cardMap[row][col]) {
+        possibilities.push([col, row + 1]);
       }
     }
   }
-
-  return complement;
+  return possibilities;
 }
 
 // Examines a single square bracket-enclosed part of a note (i.e. keyword) and returns the set of
 // card possibilities that it declares
 //
 // e.g. the note keyword `red` would return `[[0,1], [0,2], [0,3], [0,4], [0,5]]`
 // and the note keyword `red 3, blue 3` would return `[[0,3], [1,3]]`
+// and the note keyword `r,b,2,3, blue 3` would return `[[0,2], [1,2], [0,3], [1,3]]`
+// and the note keyword `r,!2,!3` would return `[[0,1], [0,4], [0,5]`
 function getPossibilitiesFromKeyword(
   variant: Variant,
   keywordPreTrim: string,
 ): Array<[number, number]> | null {
-  const possibilities: Array<[number, number]> = [];
-  const negative = keywordPreTrim.startsWith("!");
-  const keyword = negative ? keywordPreTrim.substring(1) : keywordPreTrim;
-  for (const substring of keyword.split(",")) {
-    const identity = parseIdentity(variant, substring.trim());
-    if (identity.suitIndex !== null && identity.rank !== null) {
-      // Encountered an identity item, add it
-
-      // Check that this identity is not already present in the list
-      if (
-        !possibilities.some(
-          (possibility) =>
-            possibility[0] === identity.suitIndex &&
-            possibility[1] === identity.rank,
-        )
-      ) {
-        possibilities.push([identity.suitIndex, identity.rank]);
+  const positiveIdent = [];
+  const negativeIdent = [];
+  let positiveRanks = new Set<number>();
+  let positiveSuits = new Set<number>();
+  for (const substring of keywordPreTrim.split(",")) {
+    const trimmed = substring.trim();
+    const negative = trimmed.startsWith("!");
+    const identity = parseIdentity(
+      variant,
+      (negative ? trimmed.substring(1) : trimmed).trim(),
```

Comment:
This extra `.trim()` means that there can be whitespace between the `!` and its operand. Is that desirable?

Reply:
Yes, that is the intent.